app/forms/instance-create.tsx

```diff
@@ -363,14 +367,17 @@ export function CreateInstanceForm() {
         id="boot-disk-tabs"
         className="full-width"
         // default to the project images tab if there are only project images
-        defaultValue={
-          siloImages.length > 0 ? 'silo' : projectImages.length > 0 ? 'project' : 'disk'
-        }
+        defaultValue={defaultSource}
         onValueChange={(val) => {
           setValue(
             'bootDiskSourceType',
-            val === 'silo' || val === 'project' ? 'image' : 'disk'
+            val === 'silo' ? 'siloImage' : val === 'project' ? 'projectImage' : 'disk'
```

Comment:
This could just be `setValue('bootDiskSourceType', val)` if the strings were the same. Unfortunately there is no cute way to get TS to check that you're not messing it up somehow, but we're already lacking that, so it's not getting any worse.

Reply:
Best I could do. Kind of horrible. Probably best to instead just make sure the e2es exercise all three tabs.

```diff
diff --git a/app/forms/instance-create.tsx b/app/forms/instance-create.tsx
index 7156e5ae..4c55aeee 100644
--- a/app/forms/instance-create.tsx
+++ b/app/forms/instance-create.tsx
@@ -60,6 +60,8 @@ import { links } from '~/util/links'
 import { pb } from '~/util/path-builder'
 import { GiB } from '~/util/units'
 
+type BootDiskTab = 'disk' | 'projectImage' | 'siloImage'
+
 export type InstanceCreateInput = Assign<
   // API accepts undefined but it's easier if we don't
   SetRequired<InstanceCreate, 'networkInterfaces'>,
@@ -68,7 +70,7 @@ export type InstanceCreateInput = Assign<
     disks: DiskTableItem[]
     bootDiskName: string
     bootDiskSize: number
-    bootDiskSourceType: 'disk' | 'projectImage' | 'siloImage'
+    bootDiskSourceType: BootDiskTab
     bootDiskSource: string
     userData: File | null
     // ssh keys are always specified. we do not need the undefined case
@@ -161,11 +163,11 @@ export function CreateInstanceForm() {
   const allKeys = useMemo(() => sshKeys.items.map((key) => key.id), [sshKeys])
 
   const defaultSource =
-    siloImages.length > 0 ? 'silo' : projectImages.length > 0 ? 'project' : 'disk'
+    siloImages.length > 0 ? 'siloImage' : projectImages.length > 0 ? 'projectImage' : 'disk'
 
   const defaultValues: InstanceCreateInput = {
     ...baseDefaultValues,
-    bootDiskSourceType: defaultSource === 'disk' ? 'disk' : `${defaultSource}Image`,
+    bootDiskSourceType: defaultSource,
     bootDiskSource: defaultImage?.id || '',
     sshPublicKeys: allKeys,
     // Use 2x the image size as the default boot disk size
@@ -369,10 +371,7 @@ export function CreateInstanceForm() {
         // default to the project images tab if there are only project images
         defaultValue={defaultSource}
         onValueChange={(val) => {
-          setValue(
-            'bootDiskSourceType',
-            val === 'silo' ? 'siloImage' : val === 'project' ? 'projectImage' : 'disk'
-          )
+          setValue('bootDiskSourceType', val as BootDiskTab)
           // if the user switches to the disk tab, clear the selected image;
           // if they switch to the same tab that has the defaultImage, select it
           setValue('bootDiskSource', val === defaultSource ? defaultImage?.id : '')
@@ -381,13 +380,16 @@ export function CreateInstanceForm() {
         }}
       >
         <Tabs.List aria-describedby="boot-disk">
-          <Tabs.Trigger value="silo" disabled={isSubmitting}>
+          <Tabs.Trigger value={'siloImage' satisfies BootDiskTab} disabled={isSubmitting}>
             Silo images
           </Tabs.Trigger>
-          <Tabs.Trigger value="project" disabled={isSubmitting}>
+          <Tabs.Trigger
+            value={'projectImage' satisfies BootDiskTab}
+            disabled={isSubmitting}
+          >
             Project images
           </Tabs.Trigger>
-          <Tabs.Trigger value="disk" disabled={isSubmitting}>
+          <Tabs.Trigger value={'disk' satisfies BootDiskTab} disabled={isSubmitting}>
             Existing disks
           </Tabs.Trigger>
         </Tabs.List>
@@ -398,7 +400,7 @@ export function CreateInstanceForm() {
             content="Images or disks are required to create or attach a boot disk."
           />
         )}
-        <Tabs.Content value="silo" className="space-y-4">
+        <Tabs.Content value={'siloImage' satisfies BootDiskTab} className="space-y-4">
           {siloImages.length === 0 ? (
             <div className="flex max-w-lg items-center justify-center rounded-lg border p-6 border-default">
               <EmptyMessage
@@ -415,7 +417,7 @@ export function CreateInstanceForm() {
             />
           )}
         </Tabs.Content>
-        <Tabs.Content value="project" className="space-y-4">
+        <Tabs.Content value={'projectImage' satisfies BootDiskTab} className="space-y-4">
           {projectImages.length === 0 ? (
             <div className="flex max-w-lg items-center justify-center rounded-lg border p-6 border-default">
               <EmptyMessage
@@ -435,7 +437,7 @@ export function CreateInstanceForm() {
           )}
         </Tabs.Content>
 
-        <Tabs.Content value="disk" className="space-y-4">
+        <Tabs.Content value={'disk' satisfies BootDiskTab} className="space-y-4">
           {disks.length === 0 ? (
             <div className="flex max-w-lg items-center justify-center rounded-lg border p-6 border-default">
               <EmptyMessage
```

Reply:
Eh, I liked it; folded it in. Working on the tests.